### PR TITLE
Add compatibility for DSpace 7 with the Remote Handle Resolver

### LINKED ISSUE #8362: Add compatibility for DSpace 7 with the Remote Handle Resolver
**Description**
For previous DSpace versions a handle plugin was developed to allow institutions running multiple dspace instance to setup a single central handle server able to manage all their prefix. The handle plugin project is here https://github.com/DSpace/Remote-Handle-Resolver

There are two main service to resolve an Handle Indetifier:
- `${dspace.url}/xmlui/handleresolver/resolve/${handleIdentifier}` DSpace ver < 7
- `${dspace.url}/json/hdlresolver/resolve/${handleIdentifier}` DSpace ver >= 7

and they're exclusive depending on target DSpace's version.

**To Reproduce**
Navigate to two different DSpace instances and check if anything responds at that target paths.
You should notice that only one of those will respond.

**Expected behavior**
To mantain a retrocompatibility and to gradually migrate to an unique Handle Resolver, the path should be unique for all the DSpace versions.

**Related work**
https://github.com/DSpace/DSpace/pull/8357

### PR DESCRIPTION
**Pull request by vins01-4science**
## References
* Fixes #8362

## Description
Added a new public GET Method to find and retrieve handle from DSPace.
This service is needed to centralize handle retrieving in DSpace.

The path to access this new service is `http://{dspace.url}/hdlResolver/{handle}` where `{handle}` is the placeholder for the handle Identifier.

### Request Example
`curl -i -H "Accept: application/json;charset=UTF-8" -H "Content-Type: application/json;charset=UTF-8" -X GET http://{dspace.url}/hdlResolver/{handle}`

### Input / Output Examples
Handle IDentifier | Responses (Status - Body Content)
-- | --
handleIdExample/1 | **Found** : 200 - ["http://localhost/handle/hanldeIdExample1"] / **Not Found**: 200 - null
(Empty String) | 400 - Empty Body
null | 400 - Empty Body

## Instructions for Reviewers
List of changes in this PR:
* Added a new RestController with one GET Method for retrieving handle URL using handle identifier
* Added also IntegrationTests
* Added some utility methods and POJO.

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/test/java/org/dspace/builder/ItemBuilder.java [hunk 1]
```diff
@@ -83,6 +83,10 @@ public ItemBuilder withAuthor(final String authorName) {
         return addMetadataValue(item, MetadataSchemaEnum.DC.getName(), "contributor", "author", authorName);
     }
 
+    public ItemBuilder withAuthor(final String authorName, final String authority) {
+        return addMetadataValue(item, DC.getName(), "contributor", "author", null, authorName, authority, 600);
```
Comment (pnbecker) [inline dspace-api/src/test/java/org/dspace/builder/ItemBuilder.java:87]: I'm not sure if we need this short cut as we already have a method to build an Item with an author. But after thinking about it, I have not hard opinion on this.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    @Before
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+    }
+
+    /**
+     * Verifies that any null hdlIdentifier returns a
+     * <code>HttpStatus.BAD_REQUEST</code>
```
Comment (pnbecker) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java:48]: The test does something else than the comment suggest. Can you please fix the comment? The test is checking if a handle is resolved correctly, the comment suggest it would check a null handle and expect a 400 http status code.

**Review by pnbecker — COMMENTED** [review]
The code overall looks good to me. There are small things to be fixed in two comments.

@vins01-4science  @abollini 
I have one general question: have you tested this with the RemoteHandleResolver the community offers? https://github.com/DSpace/Remote-Handle-Resolver This resolver sends request to three paths:

- BASE_URL/resolve/
- BASE_URL/listhandles/
- BASE_URL/listprefixes/

This PR does support the first URL only. Do you think you can add the other two? The code from DSpace 6 that supported this can be found here: https://github.com/DSpace/DSpace/blob/5e4aec74645ee0ef877952259347992d01353e84/dspace-jspui/src/main/java/org/dspace/app/webui/handle/HandleJSONResolver.java

File: dspace-api/src/test/java/org/dspace/builder/ItemBuilder.java (hunk 1 above)
Comment (abollini) [reply dspace-api/src/test/java/org/dspace/builder/ItemBuilder.java:87]: @vins01-4science it looks like changes to this file can be reverted, they are unrelated with the PR purpose

**Comment by abollini** [comment]
@pnbecker you are right the other two endpoints are (could) be needed to, they seems to be not used by CNRI when during a single handle redirect but maybe (probably) they are used by them in other duty talking with the local handle server. So we will add them

**Comment by vins01-4science** [comment]
@abollini @tdonohue @pnbecker 
PR updated with requested changes.
Thank you!

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * This controller is public and is useful for handle resolving,
+ * wheter a target handle identifier will be resolved into the
+ * corresponding URL (if found), otherwise will respond a null string.
+ * 
+ * @author Vincenzo Mecca (vins01-4science - vincenzo.mecca at 4science.it)
+ *
+ */
+@RestController
+@RequestMapping(path = "/{hdlService:hdlResolver|resolve|listhandles|listprefixes}/")
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:37]: Could you add a REST Contract for this new endpoint?  That way it's fully described.

I also feel this is necessary because the `@RequestMapping` looks odd here. Unless I'm misunderstanding the syntax, it appears this creates several top-level paths?  `/api/hdlResolver/`, `/api/resolve/`, etc.  Why are these not structured similar to other endpoints?  It seems like they all should be grouped under `/api/hdlResolver/_subpath_` or something similar?

Additionally, we don't usually use camel-case to define endpoints.  So, `hdlResolver` may need to become `hdlresolver` or `handleresolver`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    private HdlResolverService hdlResolverService;
+
+    /**
+     * REST GET Method used to find and retrieve the URL of a target Handle. It
+     * should return only one item, if found, else a null body value.
+     * 
+     * Generate an <code>HttpStatus.BAD_REQUEST</code> 400 Error if the handle used
+     * in path isn't valid.
+     * 
+     * The response type will be the same that is in the /xmlui/handleresolver/resolve
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:56]: We should remove references to XMLUI here... they don't make sense in DSpace 7 as the XMLUI doesn't exist anymore.  Instead of this reference, I'd recommend including an example request which shows how this method works & how it might respond

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverService.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package org.dspace.app.rest.hdlresolver;
+
+import org.dspace.core.Context;
+
+/**
+ * @author Vincenzo Mecca (vins01-4science - vincenzo.mecca at 4science.it)
+ *
+ */
+public interface HdlResolverService {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverService.java:16]: Please add JavaDocs to this interface to describe it.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverDTO.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest.hdlresolver;
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverDTO.java:8]: Why is this DTO in `dspace-server-webapp`?  It looks (to me) like this is the first DTO we are distributing in webapp directly.  Arguably, the DTO (along with the `HdlResolverService` and implementation should be in `dspace-api` instead...alongside all other DTOs.  Only the `RestController` looks like it's specific to the REST API.

It seems, to me, that this might be more appropriately moved to the existing `org.dspace.handle` package (in dspace-api), perhaps under a new subpath named `org.dspace.handle.hdlresolver`?   But, if I'm missing some reason why it should be in the REST API, let me know.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+     * This method will be run before every test as per @Before. It will initialize
+     * resources required for the tests.
+     *
+     * Other methods can be annotated with @Before here or in subclasses but no
+     * execution order is guaranteed
+     */
+    @Before
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java:44]: This method can be removed. There's no reason to override the default `setUp()` if you only call `super.setUp()`.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        getClient()
+            .perform(get(HdlResolverRestController.RESOLVE + "testHdlResolver/2"))
+            .andExpect(status().isOk()).andExpect(content().string("null"));
+        getClient()
+            .perform(get(HdlResolverRestController.LISTHANDLES + "testHdlResolver/2"))
+            .andExpect(status().isOk()).andExpect(content().string("null"));
+        getClient()
+            .perform(get(HdlResolverRestController.LISTPREFIXES + "testHdlResolver/2"))
+            .andExpect(status().isOk()).andExpect(content().string("null"));
+    }
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java:170]: The tests in this IT seem extremely minimal.  I'm not seeing any tests with differentiate the behaviors of `HDL_RESOLVER` vs `RESOLVE` vs `LISTHANDLES` vs `LISTPREFIXES`.  How should these endpoints work *differently*?  Currently all these tests just prove each endpoint works in the exact same way & returns the exact same response at all times (which would make me ask why do we have 4 endpoints here instead of one?).  Perhaps the REST Contract could help here, as I'd expect it to describe the differences...and then we need to ensure that tests exists to prove those differences.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@vins01-4science : Thanks for this PR!  I gave it a code review today (I don't have a remote handle resolver to test with, so I'm having difficulty figuring out how best to test it).  Overall the code looks good, but I have some minor suggestions inline for code cleanup/improvement.

As noted below, I also think this needs a REST Contract as it's currently unclear to me how these endpoints are supposed to work.  The tests don't really show the behavior well (see comments below) and I don't have a way to test this manually yet.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java (hunk 8 above)
Comment (vins01-4science) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java:170]: As requested by @pnbecker in this comment https://github.com/DSpace/DSpace/pull/8357#pullrequestreview-1046556703 the same rest service now responds with the same behaviour to all that endpoints.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@vins01-4science : This PR is currently causing all Integration Tests to fail.  It appears to be a Spring bean loading issue.  Here's the error:

```
java.lang.IllegalStateException: Failed to load ApplicationContext
...
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'hdlResolverRestController': Unsatisfied dependency expressed through field 'hdlResolverService'; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.dspace.handle.hdlresolver.HdlResolverService' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)}
```

**Comment by vins01-4science** [comment]
Hello @tdonohue ,

thank you for the review, now it should work.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * This controller is public and is useful for handle resolving,
+ * wheter a target handle identifier will be resolved into the
+ * corresponding URL (if found), otherwise will respond a null string.
+ * 
+ * @author Vincenzo Mecca (vins01-4science - vincenzo.mecca at 4science.it)
+ *
+ */
+@RestController
+@RequestMapping(path = "/{hdlService:hdlResolver|resolve|listhandles|listprefixes}/")
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:39]: As previously noted, this PR is missing a REST Contract or other Documentation to describe the behavior.  

Looking more closely at the code, it appears it creates 4 new endpoints outside of the `/api` path?
* `${dspace.server.url}/hdlResolver/{handle-prefix}/{handle-suffix}`
* `${dspace.server.url}/resolve/{handle-prefix}/{handle-suffix}`
* `${dspace.server.url}/listhandles/{handle-prefix}/{handle-suffix}`
* `${dspace.server.url}/listprefixes/{handle-prefix}/{handle-suffix}`

The examples of how to interact with these endpoints are lacking, as the tests are extremely basic.  They also all appear to do the same thing, which is to return the URL of whatever handle they are passed.

It's also unclear why the first endpoint is camel-cased (`hdlResolver`) instead of all lowercase (`hdlresolver`) as usual.

While I realize these are **not** traditional REST endpoints, the behavior here is odd to me & undocumented.  Could we consider adding these to our REST Contract?  Or if we feel they are better described elsewhere, maybe we could add details about these endpoints to the documentation for https://github.com/DSpace/Remote-Handle-Resolver ?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * This controller is public and is useful for handle resolving,
+ * wheter a target handle identifier will be resolved into the
+ * corresponding URL (if found), otherwise will respond a null string.
+ * 
+ * @author Vincenzo Mecca (vins01-4science - vincenzo.mecca at 4science.it)
+ *
+ */
+@RestController
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:38]: I would recommend we offer a way to disable this controller, as it seems like it's only needed for sites that use https://github.com/DSpace/Remote-Handle-Resolver. For any other repositories it's unnecessary and could bog down the site if crawlers hit it excessively.

Perhaps we could add something like:
```
@ConditionalOnProperty("handle.remote-resolver.enabled")
```

And then add a new `handle.remote-resolver.enabled = false` (by default) to the dspace.cfg alongside the other `handle.*` settings in that file?
```
# Whether to enable the DSpace handle resolver endpoints necessary for
# https://github.com/DSpace/Remote-Handle-Resolver
# Defaults to "false" which means these handle resolver endpoints are not available.
#handle.remote-resolver.enabled = false
```

I noticed we have an older configuration called `handle.hide.listhandles` which used to serve a similar purpose. We can remove that unused config and replace it with this new one.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@vins01-4science and @abollini :  Overall this looks OK, but it's still missing documentation, and I'm not seeing a way to disable it (by default it likely should be disabled).  

I also noticed one of the endpoints (`hdlResolver`) uses camel-case, when it probably should be all lowercase unless we plan to change our [REST API naming policies](https://github.com/DSpace/RestContract#on-the-naming-of-endpoints).  

I'm not able to easily test this myself (as I don't have a remote handle server available). But, I was able to verify that calls to these endpoints all return the corresponding object URL from the UI (e.g. calling http://localhost:8080/server/hdlResolver/123456789/1 returns http://localhost:4000/handle/123456789/1).  So, it appears to be working as much as I understand it.

More comments below.

Overall, I think this is working as it should, but I do think some more minor cleanup is needed.

**Comment by pnbecker** [comment]
In DSpace 6 there was no option to disable it. All it does is to return lists of handle prefixes and an endpoint to resolve a handle into an URL. It does not interfere with the other handle server. So, why do we need an option to disable it?

**Comment by tdonohue** [comment]
@pnbecker : it was possible to disable at least part of the functionality in DSpace 6 using `handle.hide.listhandles = false`.  That configuration still exists but it is unused in the new code. I'd argue we should extend that configuration to allow all endpoints to be disabled (not just `/listhandles`) simply because 95% of all sites won't need this functionality out of DSpace.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java (hunk 9 above)
Comment (vins01-4science) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:39]: > The code overall looks good to me. There are small things to be fixed in two comments.
> 
> @vins01-4science @abollini I have one general question: have you tested this with the RemoteHandleResolver the community offers? https://github.com/DSpace/Remote-Handle-Resolver This resolver sends request to three paths:
> 
> * BASE_URL/resolve/
> * BASE_URL/listhandles/
> * BASE_URL/listprefixes/
> 
> This PR does support the first URL only. Do you think you can add the other two? The code from DSpace 6 that supported this can be found here: https://github.com/DSpace/DSpace/blob/5e4aec74645ee0ef877952259347992d01353e84/dspace-jspui/src/main/java/org/dspace/app/webui/handle/HandleJSONResolver.java

Hello @tdonohue ,

the new endpoints have been suggested by @pnbecker.
Firstly the one and only endpoints was `hdlResolver`.

The behaviour is the same as in version 6, resolve an handle into a list of URLs.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java (hunk 10 above)
Comment (vins01-4science) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:39]: Done, replaced with the new property and used to conditionally create the `HdlResolver`.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java (hunk 9 above)
Comment (vins01-4science) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:41]: Hello @tdonohue , as said before, @pnbecker asked to extend the behaviour of the rest service that was responding under the path `hdlResolver` to work also with the other paths, then @abollini asked to add the paths, so here are the three paths.

For the lowercase I'll modify, but please let me know what to do with the other paths.

Thank you.

Comment (vins01-4science) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:41]: @tdonohue : Thank you for the response, so as of now to let this PR be merged I should proceed with these changes?
- rename of `hdlResolver` to `hdlresolver`;
- remove the newly introduced endpoints: `resolve`, `listhandles`, `listprefixes`.

Is there something else?

Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:41]: @vins01-4science : This PR cannot be merged until it is tested.  I'm unfortunately not able to test it as I don't have a remote handle server to use.

So, it's waiting on @pnbecker to test this with https://github.com/DSpace/Remote-Handle-Resolver.   Based on my reading of the code though, my suspicions are that we need to **keep** the `resolve`, `listhandles`, and `listprefixes` endpoints, *but they need to be fixed to act the same as they did in DSpace 6.x*.  Currently, as I've said their behavior looks to have changed.  However, I'm hoping that @pnbecker can verify whether my suspicions are correct (as it will require testing this PR with the `DSpace/Remote-Handle-Resolver` codebase.

If there is anyone at 4Science who has successfully tested this PR with the `DSpace/Remote-Handle-Resolver` code, they may also be able to answer my questions about the `resolve`, `listhandles` and `listprefixes` endpoints.

In the meantime, **yes** the `hdlResolver` should be renamed to `hdlresolver`.  That definitely needs to be fixed.

At this time, I'm uncertain how to move this PR forward until I can find a tester (so I'm hoping @pnbecker will find time to help).  I'm not comfortable with merging this without testing it, because of my questions about the `resolve`, `listhandles` and `listprefixes` endpoints.

Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:41]: hi @tdonohue ,
there was an internal misunderstand.
You are right the three endpoints have different purposes and it is required to restore their proper behavior.
More specifically, 
- the `/resolve' endpoint is needed to allow the resolution of an exact handle (the behavior implemented so far)
- the `/listprefixes` endpoint is needed to allow the handle server to discover which prefixes are managed by a specific dspace instance, as the ultimate goal of the  `DSpace/Remote-Handle-Resolve` is to allow a single handle server to serve multiple dspace instances
- the `listhandles` endpoint was required by the internal handle system interface but I'm not sure if it is really used in normal operation.

In short, this PR needs to be updated to fix the behavior of the three different endpoints.
It should be possible to test this PR also without a real handle server as you can start the handle server locally configuring it accordingly to the documentation in the `DSpace/Remote-Handle-Resolver` and access it local http interface at port 8000. From here you will be able to query your fake handle and see that the handle server will talk with the /resolve endpoint of DSpace. The /listprefixes endpoint of DSpace is instead call when the handle server start

**Comment by abollini** [comment]
@pnbecker @tdonohue is right about the need to disable completely this endpoint as done now via the `handle.remote-resolver.enabled` property. Indeed, this was possible also in previous version as the hdlresolver were implemented via a specific plugin that, beside to be configured by default https://github.com/DSpace/DSpace/blob/dspace-5.11/dspace/config/dspace.cfg#L1411 , was not really required to be registered in the dspace.cfg

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+     *          <li>Response: 200 - ["http://localhost/handle/hanldeIdExample1"]
+     *      </ul>
+     * 
+     * @param request HttpServletRequest
+     * @return One element List or <code>null</code> with status 200
+     *         <code>HttpStatus.OK</code> or <code>HttpStatus.BAD_REQUEST</code> 400
+     *         error
+     */
+    @GetMapping(
+        value = "**",
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:75]: this should be limited to `<controller>/resolve/` requests

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+                       request.getRequestURI(),
+                       MessageFormat.format("{0}/{1}/", request.getContextPath(), hdlService)
+               );
+        if (!handleResolver.isValid()) {
+            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+        } else {
+            return new ResponseEntity<>(this.resolveToURL(request, handleResolver), HttpStatus.OK);
+        }
+    }
+
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:90]: add here 
the implementation and mapping for the `<controller>/listprefixes`
https://github.com/DSpace/DSpace/blob/5e4aec74645ee0ef877952259347992d01353e84/dspace-jspui/src/main/java/org/dspace/app/webui/handle/HandleJSONResolver.java#L69-L83

and `<controller>/resolve/listhandles`
https://github.com/DSpace/DSpace/blob/5e4aec74645ee0ef877952259347992d01353e84/dspace-jspui/src/main/java/org/dspace/app/webui/handle/HandleJSONResolver.java#L85-L101

**Comment by vins01-4science** [comment]
@tdonohue @abollini : all requested changes are now resolved.
Thank you.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@vins01-4science : I re-reviewed this today and the code looks good.  I manually called each of the new endpoints, and they all work **except for `listhandles`** (which always returns `null` for me).  I cannot figure out how to get it to work.

Here's what I tried:
* Spun up backend with this PR installed and the following in my local.cfg:
  ```
  handle.remote-resolver.enabled = true
  handle.hide.listhandles = false
  ```
* Tried all endpoints manually:
    * http://localhost:8080/server/hdlresolver/123456789/1 -> correctly returns `["http://localhost:4000/handle/123456789/1"]`
    * http://localhost:8080/server/resolve/123456789/1 -> correctly returns `["http://localhost:4000/handle/123456789/1"]`
    * http://localhost:8080/server/listprefixes -> correctly returns `["123456789"]`
    * http://localhost:8080/server/listhandles -> WRONGLY RETURNS `null`
    * http://localhost:8080/server/listhandles/123456789 -> WRONGLY RETURNS `null`

I think this is mostly working, but I cannot determine how to test `listhandles`.  Your automated tests all pass, but I cannot reproduce that behavior on my local instance.  I've also verified that my `handle` database table has a large number of handles that begin with the prefix `123456789`.

**Comment by vins01-4science** [comment]
> @vins01-4science : I re-reviewed this today and the code looks good. I manually called each of the new endpoints, and they all work **except for `listhandles`** (which always returns `null` for me). I cannot figure out how to get it to work.
> 
> Here's what I tried:
> 
> * Spun up backend with this PR installed and the following in my local.cfg:
>   ```
>   handle.remote-resolver.enabled = true
>   handle.hide.listhandles = false
>   ```
> * Tried all endpoints manually:
>   
>   * http://localhost:8080/server/hdlresolver/123456789/1 -> correctly returns `["http://localhost:4000/handle/123456789/1"]`
>   * http://localhost:8080/server/resolve/123456789/1 -> correctly returns `["http://localhost:4000/handle/123456789/1"]`
>   * http://localhost:8080/server/listprefixes -> correctly returns `["123456789"]`
>   * http://localhost:8080/server/listhandles -> WRONGLY RETURNS `null`
>   * http://localhost:8080/server/listhandles/123456789 -> WRONGLY RETURNS `null`
> 
> I think this is mostly working, but I cannot determine how to test `listhandles`. Your automated tests all pass, but I cannot reproduce that behavior on my local instance. I've also verified that my `handle` database table has a large number of handles that begin with the prefix `123456789`.

@tdonohue: now all the endpoints should work correctly.
Now, you should have:
- http://localhost:8080/server/hdlresolver/123456789/1 -> returning `["http://localhost:4000/handle/123456789/1"]`
- http://localhost:8080/server/resolve/123456789/1 -> returning `["http://localhost:4000/handle/123456789/1"]`
- http://localhost:8080/server/listprefixes -> returning `["123456789"]`
- http://localhost:8080/server/listhandles -> returning error `400 - BAD REQUEST`
- http://localhost:8080/server/listhandles/123456789 -> returning `["123456789/1"]`

Thank you!

**Review by tdonohue — APPROVED** [review]
👍 Thanks @vins01-4science !  I re-tested today and all the endpoints appear to work as expected.  I'm still not able to easily test this with https://github.com/DSpace/Remote-Handle-Resolver/, but I've verified the endpoints manually.  I've also verified this code doesn't impact the existing `start-handle-server` script for the embedded Handle Server.

So, this all looks good to me.

**Comment by tdonohue** [comment]
@pnbecker : Any chance you could give this a quick review so that we can get this into 7.4?  I could really use the help in moving this forward.

**Comment by tdonohue** [comment]
Re-tested today that this has no negative impact on the embedded handle-server (or `start-handle-server` CLI tool) and it does not.  Therefore, I'm going to merge this with `1 approval` as no one else has been available to help test this, the code appears to work & it's standalone & disabled by default

**Comment by pnbecker** [comment]
I tested it again today, works great with https://github.com/DSpace/Remote-Handle-Resolver/. I updated the documentation in the [DSpace Manual](https://wiki.lyrasis.org/display/DSDOC7x/Handle.Net+Registry+Support).
